Render URLs as clickable links in timeline

MidhunSureshR opened the pull request:
Implements feature request #246.

MidhunSureshR committed:
Implement object format to represent chat messages

Every chat text message can be split into parts such as text, newline
and links.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Implement linkifier

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add tests for linkify

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Return new message format instead of message text

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Transform new message format into DOM elements

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/MessageObjectFormat.js:
```diff
@@ -0,0 +1,24 @@
+export class MessageObjectFormat {
```
I like the concept of this class, but the name is a bit implementation-detail-y. What about `MessageBody` or `MessageBodyBuilder`?

Also, can we put this and the linkifier in `src/domain/session/room/timeline` as it is specific to the timeline?

bwindels commented on src/domain/Linkifier.js:
```diff
@@ -0,0 +1,173 @@
+     * Separate string into text, newlines and add them into message object.
+     * @param {String} text 
+     */
+    _addTextToMessage(text) {
```
This doesn't seem related to linkifying, I wonder if we should move the newline handling to `MessageObjectFormat`. So for every line `MessageObjectFormat` detects, it calls the linkifier with that line to split it up. Perhaps the `Linkifier` class can even be a `linkify` function given it has so little state, where apart from the text you pass in a `callback(isLink, text)` which `MessageObjectFormat` then translates into `insertText` (if `isLink` is `false`) or `insertLink` calls internally. Wdyt?

bwindels commented on src/domain/session/room/timeline/tiles/TextTile.js:
```diff
@@ -15,15 +15,16 @@ limitations under the License.
+    get messageFormat() {
         const content = this._getContent();
-        const body = content && content.body;
+        let body = content && content.body;
```
I think the linkifier won't work if body is undefined here, so perhaps we can fall back to an empty string
```suggestion
        let body = content?.body || "";
```

bwindels commented on src/domain/session/room/timeline/tiles/TextTile.js:
```diff
@@ -15,15 +15,16 @@ limitations under the License.
 
 export class TextTile extends MessageTile {
-    get text() {
+    get messageFormat() {
```
I'd expect a format to be a format identifier like `html` or something ... not the actual message content. What about just calling this `body`?

bwindels commented on src/platform/web/ui/session/room/timeline/TextMessageView.js:
```diff
@@ -21,28 +21,30 @@ import {renderMessage} from "./common.js";
+        const children = [];
+        for (const m of value) {
+            const f = this._formatFunction[m.type];
+            const element = f({ obj: m, t: t });
```
perhaps don't create an extra object here, but just pass `m` and `t` as separate arguments? Otherwise looks great!

bwindels commented on src/platform/web/ui/session/room/timeline/TextMessageView.js:
```diff
@@ -21,28 +21,30 @@ import {renderMessage} from "./common.js";
 class BodyView extends StaticView {
+
+    get _formatFunction() {
+        return {
```
this object can probably just exist as a variable at the module level so it doesn't get recreated every time `_formatFunction` is called.

bwindels reviewed (changes requested):
This looks great, thank you!

A few comments about naming, and one about keeping the linkifier just about links, rather than also handling newlines. Other than that, looking great :+1:

MidhunSureshR committed:
Rename MessageBodyBuilder to MessageBodyBuilder

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Moved files into more appropriate locations

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Refactored Linkifier class into single function

Modified design so that linkify interacts with MessageBodyBuilder
through callbacks.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move newline handling to MessageBodyBuilder

- Also add a insert method to reflect new design.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Use MessageBodyBuilder instead of linkify

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add newline to message

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Added tests for new linkify

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Refactor loop

Don't handle last element separately

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fallback to empty string if content is undefined

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Rename messageFormat to body

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix bug in which links preceeded text

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Refactor unwanted getter into constant object

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Refactor functions to accept single parameter

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Cache result in TextTile

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/session/room/timeline/MessageBodyBuilder.js:
```diff
@@ -0,0 +1,161 @@
+    fromText(text) {
+        const components = text.split("\n");
+        components.flatMap(e => ["\n", e]).slice(1).forEach(e => {
+            if (e === "\n")
```
please use brackets for all block statement (this should really be in a style guide, sorry)

bwindels commented on src/domain/session/room/timeline/MessageBodyBuilder.js:
```diff
@@ -0,0 +1,161 @@
+
+        // Tests for links
+        "Link with host": assert => {
+            testLink(assert, "https://matrix.org");
```
can we split up the tests for links and put them in `linkify.js`?

bwindels commented on src/domain/session/room/timeline/linkify.js:
```diff
@@ -0,0 +1,12 @@
+export function linkify(text, callback) {
+    const regex = /(?:https|http|ftp):\/\/[a-zA-Z0-9:.\[\]#-]+(?:\/[^\s]*[^\s.,?!]|[^\s\u{80}-\u{10ffff}.,?!])/gui
```
define this at module scope so its doesn't get recreated on every invocation?

bwindels commented on src/domain/session/room/timeline/linkify.js:
```diff
@@ -0,0 +1,12 @@
+export function linkify(text, callback) {
+    const regex = /(?:https|http|ftp):\/\/[a-zA-Z0-9:.\[\]#-]+(?:\/[^\s]*[^\s.,?!]|[^\s\u{80}-\u{10ffff}.,?!])/gui
```
Can we have some comments explaining the different parts of the regex?
I'm also somewhat confused by:
 - the `#` in `[a-zA-Z0-9:.\[\]#-]`, hosts can't have a `#` in them AFAIK?
 - `(?:\/[^\s]*[^\s.,?!]|[^\s\u{80}-\u{10ffff}.,?!])`, `\/[^\s]` selects /* until space, `[^\s.,?!]` selects anything but whitespace and punctuation, and why do we need `[^\s\u{80}-\u{10ffff}.,?!]` again (something related to unicode but forgot why "not whitespace" doesn't cut it)? Also, what is the operator precedence of `|` here? E.g. is the left operand `[^\s]*[^\s.,?!]` or `[^\s.,?!]`?

MidhunSureshR replied:
> why do we need `[^\s\u{80}-\u{10ffff}.,?!]`

If path/fragment is absent, URL must end with ASCII character.
i.e https://matrix.org🔥 -> 🔥 is not part of URL, https://matrix.org/🔥 -> 🔥 is part of URL.
AFAIK `80-10ffff` is the non-ASCII unicode range.
>Also, what is the operator precedence of | here? E.g. is the left operand `[^\s]*[^\s.,?!]` or `[^\s.,?!]`?

The left operand is  `[^\s]*[^\s.,?!]`.

bwindels reviewed (changes requested):
Thank you for the changes!
A few more bits, we're close now :)

MidhunSureshR committed:
Add braces to all blocks

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move link tests to linkify

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Make regex module-level constant

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add test for unicode after fragment without path

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Modified regex to allow unicode in fragment

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move linkify.js to separate directory

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Put regex into separate file

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Refactor regex to improve readability

- Split regex into components
- Add informative comments

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add some helpful pointers

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Explain the reason for double escapes

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Change word to improve meaning

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/domain/session/room/timeline/linkify/regex.js:
```diff
@@ -0,0 +1,36 @@
+    https://matrix.org<smiley> - invalid
+2. Do not treat punctuation at the end as a part of the URL (.,?!)
+*/
+const urlRegex = `${scheme}${host}+(?:${additional}|${endASCII})`;
```
endASCII is basically checking the last character of the host part, right? It should have the same constraints as `host`, and additionally not accept any punctuation. This is true whether there is a path/fragment or not. So could we instead simplify to this?:
```js
const scheme = "(?:https|http|ftp):\\/\\/";
const noSpaceNorPunctuation = "[^\\s.,?!]";
const hostCharacter = "[a-zA-Z0-9:.\\[\\]-]";
// using non-consuming group here to combine two criteria for the last character
const host = `${hostCharacter}*(?=${hostCharacter})${noSpaceNorPunctuation}`;
// use sub groups so we accept just / or #, but if anything comes after it should not end with punctuation
const pathOrFragment = `(?:[\\/#](?:[^\\s]*${noSpaceNorPunctuation})?)`;
// path or fragment is optional
const urlRegex = `${scheme}${host}${pathOrFragment}?`;
```

This way the last character of the host won't accept any characters not accepted by DNS, like `https://matrix.org<`

bwindels commented on src/domain/session/room/timeline/linkify/regex.js:
```diff
@@ -0,0 +1,36 @@
+/*
```
Thanks for splitting it up, much more readable like this!

bwindels commented on src/domain/session/room/timeline/linkify/linkify.js:
```diff
@@ -0,0 +1,112 @@
+    const matches = text.matchAll(regex);
+    let curr = 0;
+    for (let match of matches) {
+        callback(text.slice(curr, match.index), false);
```
perhaps assign it to a variable so it's clear what is happening? Same below.
```suggestion
        const precedingText = text.slice(curr, match.index);
        callback(precedingText, false);
```

bwindels commented on src/domain/session/room/timeline/linkify/linkify.js:
```diff
@@ -0,0 +1,112 @@
+            testLink(assert, "https://foo.bar.com/\uD83D\uDE03");
+        },
+
+        "Link with unicode after fragment without path must linkify": assert => {
```
Great to have so many tests? Can we have a test for `https://matrix.org<` as well which seems to be accepted with the current regex (see below)?

bwindels reviewed (changes requested):
Thank you for your changes! nearly there, just a few more comments about the regex.

bwindels commented on src/domain/session/room/timeline/MessageBodyBuilder.js:
```diff
@@ -0,0 +1,96 @@
+
+    fromText(text) {
+        const components = text.split("\n");
+        components.flatMap(e => ["\n", e]).slice(1).forEach(e => {
```
flatMap is not supported by IE11, the legacy build _should_ include a polyfill for it from corejs and put it in the legacy bundle. We should double check this is actually happening before merging.

MidhunSureshR committed:
Pull callback args into variables

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Add failing test for link ending with <

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Grammar fix

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Fix broken test

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Incorporate better regex from review

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels reviewed (approved):
I think we have a winner, will just try it out before and then merge :+1: Thanks for seeing this one out :slightly_smiling_face:
